src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,427 @@
+    where
+        Base: HugrView,
+    {
+        let (inp, out) = hugr.children(root).take(2).collect_tuple().unwrap();
+        let incoming = BoundaryEdge::get_boundary_outgoing(inp, hugr).collect();
+        let outgoing = BoundaryEdge::get_boundary_incoming(out, hugr).collect();
+        Self {
+            hugr,
+            root,
+            incoming,
+            outgoing,
+            sibling_graph: OnceCell::new(),
+        }
+    }
+
+    /// Creates a new sibling subgraph.
+    ///
+    /// The incoming and outgoing edges can be expressed as either source or
+    /// target ports. If source ports are given, they will be converted to
+    /// target ports, and, in the presence of copies, the signature will be
+    /// expanded accordingly.
+    pub fn new(
+        hugr: &'g Base,
+        root: Node,
+        incoming: impl IntoIterator<Item = (Node, Port)>,
+        outgoing: impl IntoIterator<Item = (Node, Port)>,
+    ) -> Self
+    where
+        Base: HugrView,
+    {
+        let incoming = incoming
+            .into_iter()
+            .flat_map(|(n, p)| BoundaryEdge::get_boundary_edges(n, p, hugr))
+            .collect();
+        let outgoing = outgoing
+            .into_iter()
+            .flat_map(|(n, p)| BoundaryEdge::get_boundary_edges(n, p, hugr))
+            .collect();
+        Self {
+            hugr,
+            root,
+            incoming,
+            outgoing,
+            sibling_graph: OnceCell::new(),
+        }
+    }
+
+    fn get_sibling_graph(&self) -> &Subgraph<'g, Base::Portgraph> {
+        self.sibling_graph.get_or_init(|| {
+            let graph = self.hugr.portgraph();
+            let incoming = self
+                .incoming
+                .iter()
+                .copied()
+                .map(|e| e.target_port_index(graph));
+            let outgoing = self
+                .outgoing
+                .iter()
+                .copied()
+                .filter_map(|e| e.source_port_index(graph));
+            Subgraph::new_subgraph(graph, incoming.chain(outgoing))
+        })
+    }
+
+    /// An iterator over the nodes in the subgraph.
+    pub fn nodes(&self) -> impl Iterator<Item = Node> + '_
+    where
+        Base: HugrView,
+    {
+        self.get_sibling_graph().nodes_iter().flat_map(|index| {
+            let region = SiblingGraph::new(self.hugr, Node { index });
+            region.nodes().collect_vec()
+        })
+    }
+
+    /// The number of incoming and outgoing wires of the subgraph.
+    pub fn boundary_signature(&self) -> (usize, usize)
+    where
+        Base: HugrView,
+    {
+        let incoming = self.incoming.len();
+        let outgoing = self.outgoing.len();
+        (incoming, outgoing)
+    }
+
+    /// Construct a [`SimpleReplacement`] to replace `self` with `replacement`.
+    ///
+    /// `replacement` must be a hugr with DFG root and its signature must
+    /// match the signature of the subgraph.
+    ///
+    /// Panics if
```

Comment:
Note that `specification/hugr.md` defines SimpleReplace as requiring the replaced nodes to be convex. That check isn't implemented there, but you could check that here. (That might reduce the utility of allowing non-convex SiblingSubgraphs, too...)

Reply:
I've changed `create_simple_replacement` to return a `Result` and to check that the subgraph is convex.